infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntryWriter.java

```diff
@@ -54,83 +64,139 @@ public LogEntryWriter(CorfuRuntime rt, LogReplicationConfig config, LogReplicati
+
+                        long persistedTopologyConfigId =
+                            metadataMap.get(LogReplicationMetadataType.TOPOLOGY_CONFIG_ID);
+                        long persistedSnapshotStart =
+                            metadataMap.get(LogReplicationMetadataType.LAST_SNAPSHOT_STARTED);
+                        long persistedSnapshotDone =
+                            metadataMap.get(LogReplicationMetadataType.LAST_SNAPSHOT_APPLIED);
+                        long persistedBatchTs =
+                            metadataMap.get(LogReplicationMetadataType.LAST_LOG_ENTRY_PROCESSED);
+                        long persistedOpaqueEntryTs =
+                            metadataMap.get(LogReplicationMetadataType.LAST_OPAQUE_ENTRY_PROCESSED);
+
+                        long topologyConfigId = txMessage.getMetadata().getTopologyConfigID();
+                        long baseSnapshotTs = txMessage.getMetadata().getSnapshotTimestamp();
+                        long prevTs = txMessage.getMetadata().getPreviousTimestamp();
+
+                        // Validate the message metadata with the local metadata table
+                        // TODO: This validation should be performed only for the topology config id after the
+                        //  metadata table's schema is changed
+                        if (topologyConfigId != persistedTopologyConfigId || baseSnapshotTs != persistedSnapshotStart ||
+                            baseSnapshotTs != persistedSnapshotDone || prevTs != persistedBatchTs) {
+                            log.warn("Message metadata mismatch. Skip applying message {}, persistedTopologyConfigId={}," +
+                                    "persistedSnapshotStart={}, persistedSnapshotDone={}, persistedBatchTs={}",
+                                txMessage.getMetadata(), persistedTopologyConfigId, persistedSnapshotStart,
+                                persistedSnapshotDone, persistedBatchTs);
+                            throw new IllegalArgumentException("Cannot apply log entry message due to metadata mismatch");
+                        }
+
+                        // Skip Opaque entries with timestamp that are not larger than persistedOpaqueEntryTs
+                        if (opaqueEntry.getVersion() <= persistedOpaqueEntryTs) {
+                            log.trace("Skipping entry {} as it is less than the last applied opaque entry {}",
+                                opaqueEntry.getVersion(), persistedOpaqueEntryTs);
+                            return null;
+                        }
+
+                        // LAST_OPAQUE_ENTRY_PROCESSED has the timestamp of the last OpaqueEntry applied from a
+                        // batch of opaque entries received.
+                        logReplicationMetadataManager.appendUpdate(txnContext,
+                            LogReplicationMetadataType.LAST_OPAQUE_ENTRY_PROCESSED, opaqueEntry.getVersion());
```

Comment:
I think this should be at the end..after we write the opaque entry.
Else it can lead to data loss...
say we had a LR msg with D11,D12,D13. We append D11, and during D12 we update our metadata but node restarts before the data is written. Now when the node comes again and the same LR msg is received again, we will end up dropping D11,D12

Reply:
The transaction is not being committed here so a partial update of metadata without data will not occur.  This update is happening atomically with the data write.